woocommerce/woocommerce
# Enable required checkboxes in Additional Checkout Fields API

### Changes proposed in this Pull Request:

- Adds a `ValidatedCheckboxControl` component to the `@woocommerce/blocks-components` package.
- Add tests to cover `ValidatedCheckboxControl`
- Support `'required' => true` for checkbox fields in the Additional Checkout Fields API.
- Support custom error messages for required checkbox fields in the Additional Checkout Fields API.
- Bug fix in `CheckboxControl` story. (Will add `ValidatedCheckboxControl` stories in a follow up PR)



Closes #51255



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

## Front-end (shopper) testing
1. Add the following code to your site:

```php
add_action( 'woocommerce_blocks_loaded', function() {
woocommerce_register_additional_checkout_field(
			array(
				'id'       => 'first-plugin-namespace/test-required-field',
				'label'    => 'Test required field',
				'location' => 'contact',
				'required' => true, // boolean, not string
				'type'     => 'checkbox',
				'error_message' => 'Please check the box or you cannot place the order'
			)
		);
} );
```

2. Go to the Checkout block and ensure you see the `Test required field` checkbox.
3. Check it and uncheck it, verify the error message defined above shows up.
4. Try to check out while it's unchecked with the error message visible, ensure checkout cannot proceed while the box is unchecked.
5. Check the box, ensure the error disappears.
6. Reload the page, ensure the error message is not showing.
7. Try to check out while the checkbox is unchecked, but every other field is valid. Ensure the error message displays and you can't check out.
8. Check the box and ensure you can check out.

## Developer testing

1. Take the snippet above, change the value of `required` to `'false'` (as a string, not boolean).
2. Load the Checkout block, ensure the field is optional.
3. Change the value of `required` to an array, load the Checkout block, ensure you get a doing_it_wrong error (use [Query Monitor](https://en-gb.wordpress.org/plugins/query-monitor/) to see these)
4. Remove `required` completely, go to the Checkout block and ensure the field is optional.
5. Change the value of `required` to `true` (boolean), go to the Checkout block and ensure it is required, follow some steps from above, e.g. checking and unchecking the field.
6. Change the value of `error_message` to anything that isn't a string. Ensure you see a doing_it_wrong error in the Checkout block. Check the field is still required, but a generic error message is shown.
7. Change the value of `error_message` back, and then change `required` to `false`. Reload the Checkout block, ensure you see a doing_it_wrong error.



### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
